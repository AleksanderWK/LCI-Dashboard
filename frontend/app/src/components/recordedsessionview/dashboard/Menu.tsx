import React from "react";
import {
    createStyles,
    ListItemIcon,
    makeStyles,
    Menu as MUIMenu,
    MenuItem,
    MenuList,
    Theme,
    Typography
} from "@material-ui/core";
import DeleteOutlinedIcon from "@material-ui/icons/DeleteOutlined";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        menuList: {
            padding: 0,
            color: theme.palette.text.default,
            "& .MuiListItemIcon-root": {
                minWidth: 0,
                paddingRight: theme.spacing(1)
            }
        }
    })
);

interface Props {
    open: boolean;
    anchorEl: HTMLElement | null;
    onRemoveView: () => void;
    onMenuClose: () => void;
}

export default function Menu(props: Props): JSX.Element {
    const classes = useStyles();

    return (
        <MUIMenu
            className={"noDrag"}
            anchorEl={props.anchorEl}
            getContentAnchorEl={null}
            disableAutoFocusItem={true}
            disableScrollLock={true}
            keepMounted
            open={props.open}
            onClose={props.onMenuClose}
            anchorOrigin={{
                vertical: "bottom",
                horizontal: "right"
            }}
            transformOrigin={{
                vertical: "top",
                horizontal: "right"
            }}
            elevation={3}
            data-testid="dropdown-menu"
        >
            <MenuList dense={true} className={classes.menuList}>
                <MenuItem onClick={props.onRemoveView} aria-label="remove view">
                    <ListItemIcon>
                        <DeleteOutlinedIcon />
                    </ListItemIcon>
                    <Typography variant="inherit">Remove view</Typography>
                </MenuItem>
            </MenuList>
        </MUIMenu>
    );
}
