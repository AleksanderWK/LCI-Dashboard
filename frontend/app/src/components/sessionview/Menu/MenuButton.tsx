import {createStyles, makeStyles, Theme, Tooltip} from "@material-ui/core";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        buttonWrapper: {
            position: "relative",
            display: "flex",
            justifyContent: "center"
        },
        button: {
            display: "flex",
            justifyContent: "center",
            alignItems: "center",
            width: "60px",
            height: "60px",
            borderRadius: "50%",
            border: "3px solid",
            borderColor: theme.palette.primary.main,
            boxShadow: "0px 2px 5px 0px rgba(0,0,0,0.25)",
            outline: 0,
            cursor: "pointer",
            padding: 0,
            margin: 0,
            color: theme.palette.text.default,
            transition: "box-shadow 0.15s linear",
            "&:hover": {
                boxShadow: "0px 2px 5px 0px rgba(0,0,0,0.4)"
            }
        },
        tooltip: {
            marginLeft: 10
        },
        indicator: {
            position: "absolute",
            bottom: 7,
            right: 0,
            backgroundColor: "#D6D8DD",
            width: "6px",
            height: "46px",
            borderRadius: "10px 0 0 10px"
        }
    })
);

interface Props {
    children: JSX.Element;
    selected: boolean;
    tooltip: string;
    onClick: () => void;
}

export default function MenuButton(props: Props): JSX.Element {
    const classes = useStyles();

    return (
        <div className={classes.buttonWrapper}>
            <Tooltip title={props.tooltip} classes={{tooltipPlacementRight: classes.tooltip}} placement="right">
                <button className={classes.button} onClick={props.onClick}>
                    {props.children}
                </button>
            </Tooltip>
            {props.selected && <div className={classes.indicator}></div>}
        </div>
    );
}
