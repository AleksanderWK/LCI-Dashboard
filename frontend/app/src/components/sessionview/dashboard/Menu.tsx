import React from "react";
import {
    createStyles,
    ListItemIcon,
    makeStyles,
    Menu as MUIMenu,
    MenuItem,
    MenuList,
    Theme,
    Typography
} from "@material-ui/core";
import ZoomInIcon from "@material-ui/icons/ZoomIn";
import ZoomOutIcon from "@material-ui/icons/ZoomOut";
import DeleteOutlinedIcon from "@material-ui/icons/DeleteOutlined";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        menuList: {
            padding: 0,
            color: theme.palette.text.default,
            "& .MuiListItemIcon-root": {
                minWidth: 0,
                paddingRight: theme.spacing(1)
            }
        }
    })
);

interface Props {
    open: boolean;
    anchorEl: HTMLElement | null;
    isDetailedView: boolean;
    onShowMore: () => void;
    onShowLess: () => void;
    onRemoveView: () => void;
    onMenuClose: () => void;
}

/**
 * A dropdown menu for a container in the session view.
 * Has a buttons to switch type of display and remove container.
 * @param {object} props - Component props
 * @param {boolean} props.open - Whether the dropdown is open or not
 * @param {HTMLElement | null} props.anchorEl - The anchor element for the dropdown menu
 * @param {() => void} props.onShowMore - A callback function that is called when show more button is clicked
 * @param {() => void} props.onShowLess - A callback function that is called when show less button is clicked
 * @param {() => void} props.onRemoveView - A callback function that is called when remove view button is clicked
 * @param {() => void} props.onMenuClose - A callback function that is called dropdown menu is closed
 */
export default function Menu(props: Props): JSX.Element {
    const classes = useStyles();

    return (
        <MUIMenu
            className={"noDrag"}
            anchorEl={props.anchorEl}
            getContentAnchorEl={null}
            disableAutoFocusItem={true}
            disableScrollLock={true}
            keepMounted
            open={props.open}
            onClose={props.onMenuClose}
            anchorOrigin={{
                vertical: "bottom",
                horizontal: "right"
            }}
            transformOrigin={{
                vertical: "top",
                horizontal: "right"
            }}
            elevation={3}
            data-testid="dropdown-menu"
        >
            <MenuList dense={true} className={classes.menuList}>
                {props.isDetailedView ? (
                    <MenuItem onClick={props.onShowLess} aria-label="show less">
                        <ListItemIcon>
                            <ZoomOutIcon />
                        </ListItemIcon>
                        <Typography variant="inherit">Show less</Typography>
                    </MenuItem>
                ) : (
                    <MenuItem onClick={props.onShowMore} aria-label="show more">
                        <ListItemIcon>
                            <ZoomInIcon />
                        </ListItemIcon>
                        <Typography variant="inherit">Show more</Typography>
                    </MenuItem>
                )}

                <MenuItem onClick={props.onRemoveView} aria-label="remove view">
                    <ListItemIcon>
                        <DeleteOutlinedIcon />
                    </ListItemIcon>
                    <Typography variant="inherit">Remove view</Typography>
                </MenuItem>
            </MenuList>
        </MUIMenu>
    );
}
