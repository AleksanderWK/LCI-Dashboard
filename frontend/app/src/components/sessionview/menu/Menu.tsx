import React from "react";
import {createStyles, Fab, makeStyles, Theme, Tooltip} from "@material-ui/core";
import AddIcon from "@material-ui/icons/Add";
import ClassroomButton from "./ClassroomButton";
import StudentButton from "./StudentButton";
import {useRecoilValue, useSetRecoilState} from "recoil";
import {createSessionPopupOpenState} from "../../../state/popup";
import {selectedSessionIdState, sessionsState} from "../../../state/session";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        menu: {
            position: "relative",
            boxSizing: "border-box",
            padding: "30px 0",
            width: "100%",
            height: "100%",
            backgroundColor: "#E6E6E6", // Background of title bar
            overflow: "scroll",
            gridColumnStart: 1,
            gridColumnEnd: 2,
            gridRowStart: 1,
            gridRowEnd: 3,
            // Hide scroll bar
            msOverflowStyle: "none",
            scrollbarWidth: "none",
            "&::-webkit-scrollbar": {
                display: "none"
            }
        },
        wrapper: {
            display: "grid",
            gridTemplateColumns: "100%",
            gap: "20px"
        },
        hr: {
            margin: "0 auto",
            width: "40px",
            borderRadius: "2px",
            borderBottom: "1px solid #878787"
        },
        tooltip: {
            marginLeft: 10
        },
        addButton: {
            margin: "0 auto"
        },
        addIcon: {
            color: theme.palette.text.default
        }
    })
);

export default function Menu(): JSX.Element {
    const classes = useStyles();

    const setCreateSessionPopupOpen = useSetRecoilState(createSessionPopupOpenState);

    const selectedSessionId = useRecoilValue(selectedSessionIdState);

    const sessions = useRecoilValue(sessionsState);

    return (
        <div className={classes.menu}>
            <div className={classes.wrapper}>
                <ClassroomButton selected={!selectedSessionId} />

                <div className={classes.hr}></div>

                {sessions.map((session) => (
                    <StudentButton
                        key={session._id}
                        sessionId={session._id}
                        selected={session._id === selectedSessionId}
                        studentName={session.studentName}
                    />
                ))}
                <Tooltip
                    title="Create new session"
                    classes={{tooltipPlacementRight: classes.tooltip}}
                    placement="right"
                >
                    <Fab
                        size="small"
                        color="primary"
                        aria-label="create new learning session"
                        className={classes.addButton}
                        onClick={() => {
                            setCreateSessionPopupOpen(true);
                        }}
                    >
                        <AddIcon className={classes.addIcon} />
                    </Fab>
                </Tooltip>
            </div>
        </div>
    );
}
