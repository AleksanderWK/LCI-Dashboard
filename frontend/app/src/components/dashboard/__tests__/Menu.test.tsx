import React from "react";
import ReactDOM from "react-dom";
import {render, screen} from "@testing-library/react";
import Menu from "../Menu";

it("menu matches snapshot", () => {
    const {baseElement} = render(
        <Menu
            open={true}
            anchorEl={document.body}
            isDetailedView={true}
            onShowMore={() => null}
            onShowLess={() => null}
            onRemoveView={() => null}
            onMenuClose={() => null}
        />
    );
    expect(baseElement).toMatchSnapshot();
});

it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
        <Menu
            open={true}
            anchorEl={document.body}
            isDetailedView={true}
            onShowMore={() => null}
            onShowLess={() => null}
            onRemoveView={() => null}
            onMenuClose={() => null}
        />,
        div
    );
});

it("renders show less and remove view buttons", () => {
    render(
        <Menu
            open={true}
            anchorEl={document.body}
            isDetailedView={true}
            onShowMore={() => null}
            onShowLess={() => null}
            onRemoveView={() => null}
            onMenuClose={() => null}
        />
    );
    expect(screen.getAllByText("Show less").pop()).toBeInTheDocument();
    expect(screen.getAllByText("Remove view").pop()).toBeInTheDocument();
});
