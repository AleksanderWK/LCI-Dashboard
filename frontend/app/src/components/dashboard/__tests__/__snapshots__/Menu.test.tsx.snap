// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu matches snapshot 1`] = `
<body
  style="padding-right: 0px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiPopover-root"
    data-testid="dropdown-menu"
    role="presentation"
    style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px;"
  >
    <div
      aria-hidden="true"
      style="z-index: -1; position: fixed; right: 0px; bottom: 0px; top: 0px; left: 0px; background-color: transparent;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation3 MuiPaper-rounded"
      style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
      tabindex="-1"
    >
      <ul
        class="MuiList-root MuiMenu-list MuiList-padding"
        role="menu"
        tabindex="0"
      >
        <ul
          class="MuiList-root makeStyles-menuList-1 MuiList-dense MuiList-padding"
          role="menu"
          tabindex="0"
        >
          <li
            aria-disabled="false"
            aria-label="show less"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-dense MuiMenuItem-dense MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14zM7 9h5v1H7z"
                />
              </svg>
            </div>
            <span
              class="MuiTypography-root"
            >
              Show less
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </li>
          <li
            aria-disabled="false"
            aria-label="remove view"
            class="MuiButtonBase-root MuiListItem-root MuiMenuItem-root MuiMenuItem-gutters MuiListItem-dense MuiMenuItem-dense MuiListItem-gutters MuiListItem-button"
            role="menuitem"
            tabindex="-1"
          >
            <div
              class="MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 9v10H8V9h8m-1.5-6h-5l-1 1H5v2h14V4h-3.5l-1-1zM18 7H6v12c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7z"
                />
              </svg>
            </div>
            <span
              class="MuiTypography-root"
            >
              Remove view
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </li>
        </ul>
      </ul>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
